### ISSUE TITLE
Fresco and okHttp

### ISSUE THREAD

**Issue by rooton** [facebook/fresco#1513]
Hi!

I try to use OkHTTP3 with Fresco.

**Application:**

```
public class MyApplication extends Application {

       public MyApplication () {
          super();
       }

    @Override
    public void onCreate() {
        super.onCreate();

        OkHttpClient okHttpClient = new OkHttpClient();

        ImagePipelineConfig config = OkHttpImagePipelineConfigFactory
                .newBuilder(this, okHttpClient)
        .build();

        Fresco.initialize(this, config);
    }
}
```

**Gradle:**

```
compile 'com.facebook.fresco:fresco:0.14.1'
compile 'com.squareup.okhttp3:okhttp:3.4.1'
compile 'com.facebook.fresco:imagepipeline-okhttp3:0.14.1'
```

**BaseAdapter (ListView) / getView**

```
holder.simpleDrawerView= (SimpleDraweeView) inList.findViewById(R.id.simpleDrawerView);
. . .

Uri uri = Uri.parse(getItem(pos)); //img url
//for example: Uri uri = Uri.parse("http://thumbs.fullhd.pics/350x0/974560d3c39efda1ffd3163d9df4019b/"+System.currentTimeMillis());
holder.simpleDrawerView.setImageURI(uri);
```

**Active scrolling, sometimes gives an error:**

> FATAL EXCEPTION: OkHttp Dispatcher
>                                                                         java.lang.IllegalArgumentException: in == null
>                                                                             at okio.Okio.source(Okio.java:127)
>                                                                             at okio.Okio.source(Okio.java:203)
>                                                                             at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:191)
>                                                                             at okhttp3.internal.connection.RealConnection.buildConnection(RealConnection.java:173)
>                                                                             at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:114)
>                                                                             at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:193)
>                                                                             at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:129)
>                                                                             at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:98)
>                                                                             at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
>                                                                             at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:109)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
>                                                                             at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
>                                                                             at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:124)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
>                                                                             at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
>                                                                             at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:170)
>                                                                             at okhttp3.RealCall.access$100(RealCall.java:33)
>                                                                             at okhttp3.RealCall$AsyncCall.execute(RealCall.java:120)
>                                                                             at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
>                                                                             at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
>                                                                             at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
>                                                                             at java.lang.Thread.run(Thread.java:856)

**Comment by kirwan** [comment]
Your entire stack trace relates to OkHttp and Okio. I suggest asking your question to them.

Sorry I can't be more help.

**Comment by lucas34** [comment]
Hi,

I also have this issue. In my bug reports I have this following issues 
java.lang.IllegalArgumentException: in == null
java.lang.IllegalArgumentException: out == null
Stacktrace similar has above.

After long investigation, I manage to fix the issue 

Inside the OkHttpNetworkFetcher class, the method call call.enqueue may throw exception as mentioned in the documentation of OkHttp. 

So I let the function fetchWithRequest() throw exception but I wrapped the caller.
```java
 fetchWithRequest(fetchState, callback, request);
```
with 
```java
try {
   fetchWithRequest(fetchState, callback, request);
} catch (Throwable e) {
   callback.onFailure(e);
}
```

I'm catching throwable as I follow the implementation of [Okhttp adapter for retrofit](https://github.com/JakeWharton/retrofit2-rxjava2-adapter/blob/master/src/main/java/com/jakewharton/retrofit2/adapter/rxjava2/CallObservable.java), even thought they are using .execute();

That solve the issue. 
I think fresco should have this error handling to make sure the lib won't crash the app.

Lucas.

**Comment by kirwan** [comment]
Thanks for investigating this @lucas34.

It sounds as though this is a sensible, defensive change. If you'd like to create a PR, that would be fantastic.

If you do that, please remember we have two OkHttp modules (for v2 and v3) and it would be best to update both.

**Comment by oprisnik** [comment]
Fixed in 358bdc3cd6d85a016237f156d06beffcb00d35b2
